File: src/waldiez/models/Agent/Captain/CaptainAgentData.ts
/**
 * SPDX-License-Identifier: Apache-2.0
 * Copyright 2024 - 2025 Waldiez & contributors
 */
import { WaldiezCaptainAgentLibEntry } from "@waldiez/models/Agent/Captain/types";
import {
    WaldiezAgentCodeExecutionConfig,
    WaldiezAgentData,
    WaldiezAgentHumanInputMode,
    WaldiezAgentLinkedTool,
    WaldiezAgentNestedChat,
    WaldiezAgentTerminationMessageCheck,
    WaldiezAgentUpdateSystemMessage,
} from "@waldiez/models/Agent/Common";
import { WaldiezTransitionTarget } from "@waldiez/models/common/Handoff";

/**
 * Waldiez Captain Agent Data.
 * @param humanInputMode - The human input mode of the agent ("NEVER" | "ALWAYS" | "TERMINATE")
 * @param systemMessage - The system message of the agent
 * @param codeExecutionConfig - The code execution configuration of the agent
 * @param agentDefaultAutoReply - The default auto reply of the agent
 * @param maxConsecutiveAutoReply - The maximum consecutive auto reply of the agent
 * @param termination - The termination message check of the agent
 * @param modelIds - The agent's model ids
 * @param tools - The tools available to the agent
 * @param parentId - The parent id of the agent
 * @param nestedChats - The nested chats of the agent
 * @param contextVariables - The context variables of the agent
 * @param updateAgentStateBeforeReply - The update agent state before reply of the agent
 * @param afterWork - The handoff transition after work of the agent
 * @param handoffs - The handoff / edge ids (used for ordering if needed)
 * @param toolLib - The tool library of the agent
 * @param maxRound - The maximum round of the agent
 * @param maxTurns - The maximum turns of the agent
 * @see {@link WaldiezAgentData}
 * @see {@link WaldiezAgentLinkedTool}
 * @see {@link WaldiezAgentNestedChat}
 * @see {@link WaldiezAgentTerminationMessageCheck}
 * @see {@link WaldiezAgentHumanInputMode}
 * @see {@link WaldiezAgentCodeExecutionConfig}
 * @see {@link WaldiezAgentUpdateSystemMessage}
 * @see {@link WaldiezTransitionTarget}
 */
export class WaldiezAgentCaptainData extends WaldiezAgentData {
    agentLib: WaldiezCaptainAgentLibEntry[];
    toolLib: "default" | null;
    maxRound: number;
    maxTurns: number;

    constructor(
        props: {
            humanInputMode: WaldiezAgentHumanInputMode;
            systemMessage: string | null;
            codeExecutionConfig: WaldiezAgentCodeExecutionConfig;
            agentDefaultAutoReply: string | null;
            maxConsecutiveAutoReply: number | null;
            termination: WaldiezAgentTerminationMessageCheck;
            modelIds: string[];
            tools: WaldiezAgentLinkedTool[];
            parentId?: string | null;
            nestedChats: WaldiezAgentNestedChat[];
            contextVariables: Record<string, any>;
            updateAgentStateBeforeReply: WaldiezAgentUpdateSystemMessage[];
            afterWork: WaldiezTransitionTarget | null;
            handoffs: string[]; // handoff / edge ids
            agentLib: WaldiezCaptainAgentLibEntry[];
            toolLib: "default" | null;
            maxRound: number;
            maxTurns: number;
        } = {
            humanInputMode: "NEVER",
            systemMessage: null,
            codeExecutionConfig: false,
            agentDefaultAutoReply: null,
            maxConsecutiveAutoReply: null,
            termination: {
                type: "none",
                keywords: [],
                criterion: null,
                methodContent: null,
            },
            modelIds: [],
            tools: [],
            parentId: undefined,
            nestedChats: [
                {
                    messages: [],
                    triggeredBy: [],
                    condition: {
                        conditionType: "string_llm",
                        prompt: "",
                    },
                    available: {
                        type: "none",
                        value: "",
                    },
                },
            ],
            contextVariables: {},
            updateAgentStateBeforeReply: [],
            handoffs: [],
            afterWork: null,
            agentLib: [],
            toolLib: null,
            maxRound: 10,
            maxTurns: 5,
        },
    ) {
        super(props);
        this.agentLib = props.agentLib;
        this.toolLib = props.toolLib;
        this.maxRound = props.maxRound;
        this.maxTurns = props.maxTurns;
    }
}

File: waldiez/models/agents/captain/captain_agent_data.py
# SPDX-License-Identifier: Apache-2.0.
# Copyright (c) 2024 - 2025 Waldiez and contributors.
"""Waldiez captain agent data."""

from pydantic import Field
from typing_extensions import Annotated, Literal

from ..agent import WaldiezAgentData
from .captain_agent_lib_entry import WaldiezCaptainAgentLibEntry


class WaldiezCaptainAgentData(WaldiezAgentData):
    """Captain agent data class.

    The data for a captain agent.
    Extends `WaldiezAgentData`.
    Extra attributes:
    - `agent_lib`: Optional list of agent lib entries
    - `tool_lib`:
    - `max_round`: The maximum number of rounds in a group chat
    - `max_turns`: The maximum number of turns for a chat
    See the parent's docs (`WaldiezAgentData`) for the rest of the properties.
    """

    agent_lib: Annotated[
        list[WaldiezCaptainAgentLibEntry],
        Field(
            default_factory=list,
            title="Agent lib",
            description="The agent lib",
            alias="agentLib",
        ),
    ]
    tool_lib: Annotated[
        Literal["default"] | None,
        Field(
            default=None,
            title="Tool lib",
            description="Whether to use the default tool lib",
            alias="toolLib",
        ),
    ]
    max_round: Annotated[
        int,
        Field(
            default=10,
            title="Max round",
            description="The maximum number of rounds in a group chat",
            alias="maxRound",
        ),
    ]
    max_turns: Annotated[
        int,
        Field(
            default=5,
            title="Max turns",
            description="The maximum number of turns for a chat",
            alias="maxTurns",
        ),
    ]
